pkg/client/messaging/client.go

```diff
@@ -141,108 +169,189 @@ func (c *Client) Services() []string {
 }
 
 // Send sends new message based on destination options provided.
-func (c *Client) Send(msg json.RawMessage, opts ...SendMessageOpions) error {
+func (c *Client) Send(msg json.RawMessage, opts ...SendMessageOpions) (json.RawMessage, error) {
 	sendOpts := &sendMsgOpts{}
 
 	for _, opt := range opts {
 		opt(sendOpts)
 	}
 
-	if sendOpts.connectionID != "" {
-		conn, err := c.connectionLookup.GetConnectionRecord(sendOpts.connectionID)
-		if err != nil {
-			return err
-		}
+	var action messageDispatcher
 
-		return c.sendToConnection(msg, conn)
+	didCommMsg, err := prepareMessage(msg)
+	if err != nil {
+		return nil, err
 	}
 
-	if sendOpts.theirDID != "" {
-		conn, err := c.getConnectionByTheirDID(sendOpts.theirDID)
-
-		if err != nil && err != errConnForDIDNotFound {
-			return err
-		}
+	switch {
+	case sendOpts.connectionID != "":
+		action, err = c.sendToConnection(didCommMsg, sendOpts.connectionID)
+	case sendOpts.theirDID != "":
+		action, err = c.sendToTheirDID(didCommMsg, sendOpts.theirDID)
+	case sendOpts.destination != nil:
+		action, err = c.sendToDestination(didCommMsg, sendOpts.destination)
+	default:
+		return nil, fmt.Errorf(errMsgDestinationMissing)
+	}
 
-		if conn != nil {
-			return c.sendToConnection(msg, conn)
-		}
+	if err != nil {
+		return nil, err
 	}
 
-	return c.sendToDestination(msg, sendOpts)
+	return c.sendAndWaitForReply(action, didCommMsg.ID(), sendOpts.responseMsgType, sendOpts.timeout)
 }
 
 // Reply sends reply to existing message.
-func (c *Client) Reply(msg json.RawMessage, msgID string, startNewThread bool) error {
-	didCommMsg, err := service.ParseDIDCommMsgMap(msg)
+func (c *Client) Reply(msg json.RawMessage, msgID string, startNewThread bool,
+	waitForResponse string, timeout time.Duration) (json.RawMessage, error) {
```

Comment:
you should use a context not a time.Duration.

Reply:
ok, but JS & REST users will still be using timeout.

Reply:
That's okay in those layers. In a Go API, it needs to be a context.

Reply:
done